website/dynamo.py

```diff
@@ -1566,3 +1647,139 @@ def make_predicate(obj):
+        type_dict = {}
+        for k, v in validator_dict.items():
+            if k is str:
+                type_dict[Validator.ensure(k)] = Validator.ensure(v)
+            elif isinstance(k, str):
+                type_dict[k] = Validator.ensure(v)
+            else:
+                raise ValueError(f'Key values should be of type str or instances of string. {
+                                 k} does not comply with that')
+        return type_dict
+
+    def is_valid(self, value):
+        ...
+
+    def __str__(self, value):
+        ...
+
+
+class Any(Validator):
+    """Validator which allows any type."""
+
+    def is_valid(self, value):
+        return True
+
+    def __str__(self):
+        return 'any value'
+
+
+class InstanceOf(Validator):
+    """Validator which checks if a value is an instance of a type."""
+
+    def __init__(self, type):
+        self.type = type
+
+    def is_valid(self, value):
+        return isinstance(value, self.type)
+
+    def __str__(self):
+        return f'instance of {self.type}'
+
+
+class Predicate(Validator):
+    """Validator which calls an arbitrary callback."""
+
+    def __init__(self, fn):
+        self.fn = fn
+
+    def is_valid(self, value):
+        return self.fn(value)
+
+    def __str__(self):
+        return f'matches {self.fn}'
+
+
+class Optional(Validator):
+    """Validator which matches either None or an inner validator."""
+
+    def __init__(self, inner):
+        self.inner = Validator.ensure(inner)
+
+    def is_valid(self, value):
+        return value is None or self.inner.is_valid(value)
+
+    def __str__(self):
+        return f'optional {self.inner}'
+
+
+class SetOf(Validator):
+    """Validator which matches a set matching inner validators."""
+
+    def __init__(self, inner):
+        self.inner = Validator.ensure(inner)
+
+    def is_valid(self, value):
+        return isinstance(value, set) and all(self.inner.is_valid(x) for x in value)
+
+    def __str__(self):
+        return f'set of {self.inner}'
+
+
+class ListOf(Validator):
+    """Validator which matches a list matching inner validators."""
+
+    def __init__(self, inner):
+        self.inner = Validator.ensure(inner)
+
+    def is_valid(self, value):
+        return isinstance(value, list) and all(self.inner.is_valid(x) for x in value)
+
+    def __str__(self):
+        return f'list of {self.inner}'
+
+
+class RecordOf(Validator):
+    """Validator which matches a record with inner validators."""
+
+    def __init__(self, inner):
+        self.inner = Validator.ensure_all(inner)
+
+    def is_valid(self, value):
+        if not isinstance(value, dict):
+            return False
+        first_type = list(self.inner.keys())[0]
+        if isinstance(first_type, InstanceOf):
+            return all(first_type.is_valid(k) and self.inner.get(first_type).is_valid(v) for k, v in value.items())
+        else:
+            return all(validator.is_valid(value.get(key)) for key, validator in self.inner.items())
```

Comment:
Rico, what do you think about this? I changed that you now can specify `str` in the left-hand side of the validator dictionary. This way we can be a little more generic in our type checks, like this one:

```
'sorted_adventures': RecordOf({
    str: ListOf(RecordOf({
        'name': str,
        'from_teacher': bool
}))
```

It's important to note that you can only use `str` and no other type, since that would go against Dynamo naming rules either way. This is just a way to relax the type checking a little bit without needing to use `dict`.

Reply:
A "record" (a set of known fields, with known and potentially different types for every value) is a different kind of type than a "map" (can be any field, but the types are usually the same).

Examples:

```py
person = { 
  "name": "Bertus",
  "age": "67"
}
  
favorite_restaurant_in_city = {
  "Oegstgeest": "Suzie's & Van Aken",
   "Amsterdam": "Ron's Gastropub",
}
```

In Python, you might use a `dict` to represent both of these, but the type-of-type is very different! 

They are:

```py
person :: { "name" -> str, "age" -> int }

# but
favorite_restaurant_in_city :: { str - > str }
```

So I wouldn't overload `RecordOf` in this way, because it looks like it will just get confusing. Also, it would allow you to construct this... now what does that mean?

```py
RecordOf({
  str: str,
  "age": int
})
```

I would make a `MapOf(str, str)` or `DictOf(str, str)` or something.

But you have to explain the differences very clearly, because I can see if even you are confused about it, there is a good chance other people will also not realize the difference.

Reply:
> I would make a MapOf(str, str) or DictOf(str, str) or something.

Hmm I think this would make sense! In essence what I'm trying to do is relax the checking a bit, so another name for that type of validation makes a ton of sense.

> So I wouldn't overload RecordOf in this way, because it looks like it will just get confusing. Also, it would allow you to construct this... now what does that mean?
> ```python
> RecordOf({
>  str: str,
>  "age": int
> })
> ```

I thought of that! If you use `str` in the left hand side, then you can't use a named field (and all other classes aren't allowed)

https://github.com/hedyorg/hedy/blob/dd20d554e9d2485dfbc038b2d3ba59fd94f49a5a/website/dynamo.py#L1677-L1680

> But you have to explain the differences very clearly, because I can see if even you are confused about it, there is a good chance other people will also not realize the difference

Yes, I need to add something to the wiki and to the comments.